addons/hw_drivers/connection_manager.py

```diff
@@ -12,52 +12,78 @@
-            return
+        self.running = True
+        while self.running:
+            if self._should_fetch_pairing_code():
+                if time.monotonic() > self.pairing_code_expires:
+                    self._refresh_pairing_code()
+                else:
+                    self._poll_pairing_result()
+            else:
+                self.pairing_code = False
+                self.pairing_uuid = False
+                self.pairing_code_expires = 0
+            time.sleep(10)
 
-        if not helpers.get_odoo_server_url():
-            end_time = time.monotonic() + PAIRING_CODE_TIMEOUT_SECONDS
-            while time.monotonic() < end_time:
-                self._connect_box()
-                time.sleep(10)
-            self.pairing_code = False
-            self.pairing_uuid = False
-
-    def _connect_box(self):
-        if not helpers.get_ip() or (platform.system() == 'Linux' and wifi.is_access_point()):
-            return
+    def _should_fetch_pairing_code(self):
+        return not helpers.get_odoo_server_url() and helpers.get_ip() and not (platform.system() == 'Linux' and wifi.is_access_point())
 
+    def _call_iot_proxy(self, pairing_code, pairing_uuid):
         data = {
             'jsonrpc': 2.0,
             'params': {
-                'pairing_code': self.pairing_code,
-                'pairing_uuid': self.pairing_uuid,
+                'pairing_code': pairing_code,
+                'pairing_uuid': pairing_uuid,
             }
         }
 
         try:
-            requests.packages.urllib3.disable_warnings()
             req = requests.post(
                 'https://iot-proxy.odoo.com/odoo-enterprise/iot/connect-box', json=data, verify=False, timeout=5
             )
-            result = req.json().get('result', {})
-            if all(key in result for key in ['pairing_code', 'pairing_uuid']):
-                self.pairing_code = result['pairing_code']
-                self.pairing_uuid = result['pairing_uuid']
-            elif all(key in result for key in ['url', 'token', 'db_uuid', 'enterprise_code']):
-                self._connect_to_server(result['url'], result['token'], result['db_uuid'], result['enterprise_code'])
+            return req.json().get('result', {})
         except Exception:
             _logger.exception('Could not reach iot-proxy.odoo.com')
+            return {}
+
+    def _poll_pairing_result(self):
+        result = self._call_iot_proxy(self.pairing_code, self.pairing_uuid)
+        
+        if all(key in result for key in ['url', 'token', 'db_uuid', 'enterprise_code']):
+            self._connect_to_server(result['url'], result['token'], result['db_uuid'], result['enterprise_code'])
+
+    def _refresh_pairing_code(self):
+        if self.pairing_code_count == MAXIMUM_NUMBER_OF_CODES:
+            self.running = False
+            self.pairing_code = False
+            self.pairing_uuid = False
+            return
```

Comment:
```suggestion
```
can be removed if this logic check is moved to `_should_fetch_pairing_code` as we already set these variables to false in `run` method